hypershift-operator/controllers/machinedeployment/controller.go

```diff
@@ -0,0 +1,225 @@
+package machinedeployment
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/openshift/hypershift/hypershift-operator/controllers/machinedeployment/utils"
+	"github.com/openshift/hypershift/support/awsclient"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	// These expose compute information based on the providerSpec input.
+	// This is needed by the autoscaler to foresee upcoming capacity when scaling from zero.
+	// https://github.com/openshift/enhancements/pull/186
+	cpuKey       = "machine.openshift.io/vCPU"
+	memoryKey    = "machine.openshift.io/memoryMb"
+	gpuKey       = "machine.openshift.io/GPU"
+	labelsKey    = "capacity.cluster-autoscaler.kubernetes.io/labels"
+	archLabelKey = "kubernetes.io/arch"
+
+	controlPlaneNamespaceLabel = "hypershift.openshift.io/hosted-control-plane"
+	nodePoolAnnotation         = "hypershift.openshift.io/nodePool"
+)
+
+// Reconciler reconciles MachineDeployments for scale-from-zero autoscaling.
+type Reconciler struct {
+	Client             client.Client
+	CredentialsFile    string
+	RegionCache        awsclient.RegionCache
+	InstanceTypesCache InstanceTypesCache
+
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// SetupWithManager creates a new controller for a manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.MachineDeployment{}, builder.WithPredicates(predicate.NewPredicateFuncs(r.hasNodePoolAnnotation))).
```

Comment:
It is possible that a HC has multiple nodepools, some nodepools don't have auto-scaling enabled, shall we filter only on the machinedeployment where the corresponding nodepool enabled auto-scaling ?